### ISSUE TITLE
Allow newlines after block open (or not)

### ISSUE THREAD

**Issue by vaughnkoch** [astral-sh/ruff#9745]
Hi, I'm opening this issue per a suggestion from @MichaReiser, based on this closed issue: https://github.com/astral-sh/ruff/issues/8893.

The formatting question is whether Ruff should automatically remove a blank line after a block definition:

```
def some_func():

    some_more_code()
```

to 

```
def some_func():
    some_more_code()
```

The alternative is "allow zero or one" newlines, i.e. not to remove the newline if it exists, but also not to automatically add one. This is the approach that Black chose in their 2024 preview style. Full Black discussion here: https://github.com/psf/black/issues/4043.

**Comment by ejovo13** [comment]
I second this. After exploring integrating ruff into my team's workflow as a replacement for black, isort, pydocstyle, etc., this is something I'm not willing to compromise on. Additionally, it was the first incompatibility with black that I discovered during testing. 

Excellent project though, I look forward to using it in the future.

**Comment by cbandera** [comment]
Yes I can second this request as well. While investigating the transition to ruff, this inconsistency to black is causing many changes across our code base.

**Comment by MichaReiser** [comment]
> Yes I can second this request as well. While investigating the transition to ruff, this inconsistency to black is causing many changes across our code base.

Thanks for reporting. Do you mainly mind that there are many differences or did you prefer black's formatting? Would you be able to share some examples where you preferred black's formatting over ruff's?

**Comment by cbandera** [comment]
> > Yes I can second this request as well. While investigating the transition to ruff, this inconsistency to black is causing many changes across our code base.
> 
> Thanks for reporting. Do you mainly mind that there are many differences or did you prefer black's formatting? Would you be able to share some examples where you preferred black's formatting over ruff's?

For me it's mainly about the diff that is created. Working in a big monorepo, it's always easier for us toolies to roll out changes if we don't need to touch the other devs code. Having read that `ruff` would be a "drop-in" replacement for `black`, my expectation was that I would see no diff at all.

Having said that and after looking at the [known deviations](https://docs.astral.sh/ruff/formatter/black/) I can't say that I don't agree with ruff's design decisions though...

**Comment by chicheng** [comment]
Hi, yes please add an option to preserve up to 1 empty line.

As empty lines used for separating/grouping lines, the first line(s) stay closer to block head but not the remaining parts feel like it's part of the block head (since empty line inside block content is inevitable), e.g.

```
block head
   part of block head

   block content
   block content
   block content

   block content
   block content
```

so I prefer to read as

```
block head

   block content
   block content
   block content

   block content
   block content
```

Personally I always insert an empty line at the beginning of the block, except 2 scenarios:

```python
class B(A):

    def single_line(self):    # an empty line after `class B`, because single_line() is not closer to B than other methods
        return something    # Two exceptions that no empty line after `def instance_methods():`: 1. one liner, otherwises a few one lines consecutively will feel strange as there are equal space between block head and content and next block head and content

    def override_super(self):
        super().override_super()    # 2. `super()...` call (easier to tell whether non-one-liner class method called super or not? just personal preference probably)

        remaining_codes = ...

    def function(
        self, data: torch.Tensor | tuple[torch.Tensor, ...], other_argument: int
    ) -> torch.Tensor | tuple[torch.Tensor, ...]:
        something = do_something(data)    # An empty line above please, this line is part of method body, not head. borrowed from black issue comment

        return something
```

**Comment by alechouse97** [comment]
Any updates on this? This is one the last remaining issues for ruff to finally replace black in my workflow.

**Comment by michaelhball** [comment]
Also checking in to say the same as above, I think this is our last remaining "issue" with ruff.

**Comment by dcnadler** [comment]
> Yes I can second this request as well. While investigating the transition to ruff, this inconsistency to black is causing many changes across our code base.

Agreed, this is what's blocking us from adopting Ruff. Too many diffs - not a "drop-in".

Looking forward to updates, thanks!

**Comment by thclark** [comment]
From the perspective of someone with dyslexia; this really makes code harder to read - I need whitespace around the block to help parse what's in it.

**Comment by plynch-chwy** [comment]
I was looking to make the switch to ruff and use the handful of ignored linting rules we use with pre-commit. After a couple of hours I only just found that this is an option I can't ignore or set as `unfixable`? It's kinda unfortunate that this issue is still open for almost a year as I was excited to switch.

**Comment by hankertrix** [comment]
Second this change. I only adopted black as a formatter after they implemented this change as I find it very difficult to read code without the extra newline. This is especially so when the start of the block is very long, and there's a comment right underneath the block.

For example:
```python
if long_boolean_variable_name and another_long_conditional:
    # A pretty long comment documenting what the block does
    print("Do stuff")
```

I find this to be much easier to read:
```python
if long_boolean_variable_name and another_long_conditional:

    # A pretty long comment documenting what the block does
    print("Do stuff")
```

For formatters in other languages that don't support this extra new line, I use a hack to force a new line:
```javascript
if (long_boolean_variable_name && another_long_conditional) {
    //

    // A pretty long comment documenting what the block does
    console.log("Do stuff")
}
```

Since the Pylsp formatter and black already allow an extra new line, I don't bother doing this hack for my Python code. It would be great to have Ruff support this.

**Comment by laymonage** [comment]
Hi, what's the status on this? If I'm not mistaken, Black already made this stable in 2024.

I see that this was discussed for https://github.com/astral-sh/ruff/issues/8678 in #8893. Then it was briefly mentioned in the plans for the 2025 style guide in https://github.com/astral-sh/ruff/issues/13371#issuecomment-2498529767, but it wasn't planned.

We've switched from black to ruff and loving it, but this issue prevents us from formatting Django migrations without the unnecessary noise of removing empty newlines after the `class` header that Black leaves as-is.

**Comment by dduarteo** [comment]
Was excited to start moving all of our codebases to ruff but discovered this issue with the blank line after func definitions.  I second that it makes code much easier to read.

Hopefully this gets implemented in the near future!  For now back to black.

**Comment by vitaliyf** [comment]
One more datapoint: Django generates migrations/*.py files that look like:

```
class Migration(migrations.Migration):

    dependencies = [
        ...
```

Current black-25.1.0 has no opinion on this (does not add nor remove the line break), but ruff-0.9.9 removes it.

**Comment by hauntsaninja** [comment]
We recently switched from black to ruff format at work and this issue caused many thousands of lines of diff, very little of which was an improvement and a lot of which imo hurt readability. This is currently a top 3 upvoted ruff formatter issue, so fingers crossed

**Comment by hauntsaninja** [comment]
I think the ones where it most often feels bad to me involve functions with type annotations and no docstring. There are some other cases involving dense if statements, especially nested ones or ones with comments. And a longer tail of more debatable cases, e.g. where it improves symmetry. A few not particularly cherry picked examples here: https://play.ruff.rs/bd268928-cf4e-458a-b41d-a1f163cb6347?secondary=Format

**Comment by hauntsaninja** [comment]
I'm not sure! In TypeScript and Rust, if your first statement is an assignment, the `let` distinguishes it from function parameters (it's effectively additional indent and syntax highlighting). Whereas in Python with a return type annotation it's a largely undifferentiated mass. I guess return type inference in TypeScript helps too. And I guess kwargs may mean it's more common to have functions with long signatures in Python.

I'm not sure I quite get the thing you mention about cutting, many functions have empty lines in them that aren't at code block open that we don't remove. Function signatures are read more often than function implementations, so visual distinction does at least align with a difference in use. I do agree that removing the blank line feels like a cut for one liners though.

**Comment by shivan2418** [comment]
i have the same problem. 
I was really surprised that you could not change the config the formatter more. I also want my blank lines.

**Comment by grizzlydevil** [comment]
I was testing Ruff for our project. It's super quick but the change from black to this is a bit of a deal breaker as people in the project are used to these blank lines they improve readability. Also probably we're used to it. We'll see what others decide. But this change would be very appreciated.

Thank you.

**Comment by polsanlorenzo** [comment]
Like others here, we were really excited to implement Ruff in our project due to its impressive performance, but this issue is kind of stopping us from making the move. Hopefully, this change gets implemented soon.

**Comment by rynskyi** [comment]
Same here - this is the only thing keeping us from switching from back to ruff

**Comment by MichaReiser** [comment]
I'm sympathetic to allowing a blank line after function headers. I don't love it because it trades consistency with readability but I think it's a net positive until we can come up with something better.

I'd prefer to keep this specific to functions for now, which seems to be the main concern for most. Or is this incorrect?

I'd be open to review a PR that makes this change under preview.

**Comment by guyboltonking** [comment]
> I'd prefer to keep this specific to functions for now, which seems to be the main concern for most. Or is this incorrect?

I think (and know in my case) that it applies to all blocks; I have a lot of code that has blank lines after conditionals, loops, and other constructs.

**Comment by alechouse97** [comment]
Really excited to see some traction on this! We primarily have blank lines after conditional or loop statements. Classes and functions, etc. usually have docstrings which ruff allows a new line after.

**Comment by dduarteo** [comment]
We are also in the blank line after all blocks of code camp (at least the option as does black).
Excited to see this happen and move all of our stuff to ruff!

**Comment by nonoash** [comment]
Bumping this. Same situation as others concerning the improved readability and allowing Blank lines at the start of a block. 

That is the only diff I have with Black and unfortunately prevent me from switching, much to my regret.
I hope Astral will work on their improved heuristic and adopts Black's formatting soon.

Do not forget to upvote OP

**Comment by gagaga67** [comment]
Personally, I don't want a blank line after the block like black does. This behavior is what made me choose ruff.

Why argue about what is best? It is enough to make a setting for this. Then some people will be able to add an empty line, and some will not. But forcing someone to accept one side is not a very good solution.

These are tools that should help in writing code, and not dictate what and how to do.

**Comment by MichaReiser** [comment]
@ntBre should we close this now that #21110 has merged. (I'm leaning towards yes to get a better sense in which cases where we should allow empty lines and why)

**Comment by laymonage** [comment]
@MichaReiser @ntBre What about newlines after class headers? If I'm not mistaken, #21110 only covers functions. This issue covers blocks in general – should a new issue be opened for class headers (and maybe for each header), or should this be reopened?

There's a clear demand for it, at least for classes: #8566, https://github.com/astral-sh/ruff/issues/9745#issuecomment-2023056948, https://github.com/astral-sh/ruff/issues/9745#issuecomment-2662810468, https://github.com/astral-sh/ruff/issues/9745#issuecomment-2695173158, https://github.com/astral-sh/ruff/issues/9745#issuecomment-2876498730.

Just to reiterate, our main use case is for Django migration files. It might be anecdotal but Django has a big user base, so I think it's likely there are other people wanting the same.

**Comment by MichaReiser** [comment]
I suggest opening a new issue specifically for class headers so that we get clear feedback on what's most important)
